File: app/src/main/java/com/tuhoang/pocketmind/utils/PrefsManager.java
package com.tuhoang.pocketmind.utils;

import android.content.Context;
import android.content.SharedPreferences;

import androidx.annotation.NonNull;

public class PrefsManager {

    private static volatile PrefsManager instance;
    private final SharedPreferences prefs;

    public static final String PREF_SETTINGS = "PREF_SETTINGS";
    public static final String PREF_THEME = "PREF_THEME";
    public static final String PREF_NOTIFICATIONS = "PREF_NOTIFICATIONS";
    public static final String PREF_CURRENCY = "PREF_CURRENCY";
    public static final String PREF_WORKER_URL = "PREF_WORKER_URL";
    public static final String PREF_OPENAI_API_KEY = "PREF_OPENAI_API_KEY";
    public static final String PREF_OPENAI_MODEL = "PREF_OPENAI_MODEL";
    public static final String PREF_AI_TRANSLATION_ENABLED = "PREF_AI_TRANSLATION_ENABLED";

    public static final String PREF_CAMERA_ENABLED = "PREF_CAMERA_ENABLED";
    public static final String PREF_STORAGE_ENABLED = "PREF_STORAGE_ENABLED";
    public static final String PREF_MIC_ENABLED = "PREF_MIC_ENABLED";

    private PrefsManager(@NonNull Context context) {
        prefs = context
                .getApplicationContext()
                .getSharedPreferences(PREF_SETTINGS, Context.MODE_PRIVATE);
    }

    /**
     * Init once inside Application class
     */
    public static void init(@NonNull Context context) {
        if (instance == null) {
            synchronized (PrefsManager.class) {
                if (instance == null) {
                    instance = new PrefsManager(context);
                }
            }
        }
    }

    public static PrefsManager getInstance() {
        if (instance == null) {
            throw new IllegalStateException(
                    "PrefsManager not initialized. Call PrefsManager.init() in Application.onCreate()");
        }
        return instance;
    }

    // =====================
    // THEME
    // =====================

    public int getTheme(int defaultTheme) {
        return prefs.getInt(PREF_THEME, defaultTheme);
    }

    public void setTheme(int themeMode) {
        prefs.edit()
                .putInt(PREF_THEME, themeMode)
                .apply();
    }

    // =====================
    // CURRENCY
    // =====================

    public String getCurrency(String defaultCurrency) {
        return prefs.getString(PREF_CURRENCY, defaultCurrency);
    }

    public void setCurrency(String currency) {
        prefs.edit()
                .putString(PREF_CURRENCY, currency)
                .apply();
    }

    // =====================
    // WORKER URL
    // =====================

    public String getWorkerUrl() {
        return prefs.getString(PREF_WORKER_URL, "");
    }

    public void setWorkerUrl(String url) {
        prefs.edit()
                .putString(PREF_WORKER_URL, url)
                .apply();
    }

    // =====================
    // PERMISSIONS FLAGS
    // =====================

    public boolean isCameraEnabled(boolean defaultValue) {
        return prefs.getBoolean(PREF_CAMERA_ENABLED, defaultValue);
    }

    public void setCameraEnabled(boolean enabled) {
        prefs.edit()
                .putBoolean(PREF_CAMERA_ENABLED, enabled)
                .apply();
    }

    public boolean isStorageEnabled(boolean defaultValue) {
        return prefs.getBoolean(PREF_STORAGE_ENABLED, defaultValue);
    }

    public void setStorageEnabled(boolean enabled) {
        prefs.edit()
                .putBoolean(PREF_STORAGE_ENABLED, enabled)
                .apply();
    }

    public boolean isMicEnabled(boolean defaultValue) {
        return prefs.getBoolean(PREF_MIC_ENABLED, defaultValue);
    }

    public void setMicEnabled(boolean enabled) {
        prefs.edit()
                .putBoolean(PREF_MIC_ENABLED, enabled)
                .apply();
    }

    // =====================
    // OPENAI
    // =====================

    public String getOpenAiApiKey() {
        return prefs.getString(PREF_OPENAI_API_KEY, "");
    }

    public void setOpenAiApiKey(String key) {
        prefs.edit()
                .putString(PREF_OPENAI_API_KEY, key)
                .apply();
    }

    public String getOpenAiModel(String defaultModel) {
        return prefs.getString(PREF_OPENAI_MODEL, defaultModel);
    }

    public void setOpenAiModel(String model) {
        prefs.edit()
                .putString(PREF_OPENAI_MODEL, model)
                .apply();
    }

    // =====================
    // AI TRANSLATION
    // =====================

    public boolean isAiTranslationEnabled() {
        return prefs.getBoolean(PREF_AI_TRANSLATION_ENABLED, false);
    }

    public void setAiTranslationEnabled(boolean enabled) {
        prefs.edit()
                .putBoolean(PREF_AI_TRANSLATION_ENABLED, enabled)
                .apply();
    }

}

File: app/src/main/kotlin/com/tuhoang/pocketmind/utils/PrefsManager.kt
package com.tuhoang.pocketmind.utils

import android.content.Context
import android.content.SharedPreferences

import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

class PrefsManager private constructor(context: Context) {

    private val prefs: SharedPreferences = context.applicationContext
        .getSharedPreferences(PREF_SETTINGS, Context.MODE_PRIVATE)

    private val _currencyFlow = MutableStateFlow(DEFAULT_CURRENCY)

    init {
        _currencyFlow.value = getCurrency()
    }

    companion object {
        const val PREF_SETTINGS = "PREF_SETTINGS"
        const val PREF_THEME = "PREF_THEME"
        const val PREF_CURRENCY = "PREF_CURRENCY"
        const val PREF_CUSTOM_CATEGORIES = "PREF_CUSTOM_CATEGORIES"
        const val DEFAULT_CURRENCY = "VND"
        const val PREF_WORKER_URL = "PREF_WORKER_URL"
        const val PREF_OPENAI_API_KEY = "PREF_OPENAI_API_KEY"
        const val PREF_OPENAI_MODEL = "PREF_OPENAI_MODEL"
        const val PREF_AI_TRANSLATION_ENABLED = "PREF_AI_TRANSLATION_ENABLED"
        const val PREF_CAMERA_ENABLED = "PREF_CAMERA_ENABLED"
        const val PREF_STORAGE_ENABLED = "PREF_STORAGE_ENABLED"
        const val PREF_MIC_ENABLED = "PREF_MIC_ENABLED"
        const val PREF_ONBOARDING_COMPLETE = "PREF_ONBOARDING_COMPLETE"
        const val PREF_BIOMETRIC_ENABLED = "PREF_BIOMETRIC_ENABLED"
        const val PREF_MONTHLY_BUDGET = "PREF_MONTHLY_BUDGET"
        const val PREF_DYNAMIC_COLOR = "PREF_DYNAMIC_COLOR"
        const val PREF_HAPTIC_ENABLED = "PREF_HAPTIC_ENABLED"

        @Volatile
        private var instance: PrefsManager? = null

        fun init(context: Context) {
            if (instance == null) {
                synchronized(this) {
                    if (instance == null) {
                        instance = PrefsManager(context)
                    }
                }
            }
        }

        fun getInstance(): PrefsManager =
            instance ?: error("PrefsManager not initialized. Call PrefsManager.init() in Application.onCreate()")
    }

    fun getTheme(defaultTheme: Int): Int = prefs.getInt(PREF_THEME, defaultTheme)
    fun setTheme(themeMode: Int) = prefs.edit().putInt(PREF_THEME, themeMode).apply()

    fun getCurrency(defaultCurrency: String = DEFAULT_CURRENCY): String =
        prefs.getString(PREF_CURRENCY, defaultCurrency) ?: defaultCurrency

    fun currencyFlow(): StateFlow<String> = _currencyFlow.asStateFlow()

    fun setCurrency(currency: String) {
        prefs.edit().putString(PREF_CURRENCY, currency).apply()
        _currencyFlow.value = currency
    }

    fun getCustomCategories(): List<String> {
        val raw = prefs.getString(PREF_CUSTOM_CATEGORIES, "") ?: ""
        return raw.split("|").map { it.trim() }.filter { it.isNotEmpty() }
    }

    fun setCustomCategories(categories: List<String>) {
        prefs.edit().putString(PREF_CUSTOM_CATEGORIES, categories.joinToString("|")).apply()
    }

    fun getWorkerUrl(): String = prefs.getString(PREF_WORKER_URL, "") ?: ""
    fun setWorkerUrl(url: String) = prefs.edit().putString(PREF_WORKER_URL, url).apply()

    fun isCameraEnabled(defaultValue: Boolean): Boolean =
        prefs.getBoolean(PREF_CAMERA_ENABLED, defaultValue)

    fun setCameraEnabled(enabled: Boolean) =
        prefs.edit().putBoolean(PREF_CAMERA_ENABLED, enabled).apply()

    fun isStorageEnabled(defaultValue: Boolean): Boolean =
        prefs.getBoolean(PREF_STORAGE_ENABLED, defaultValue)

    fun setStorageEnabled(enabled: Boolean) =
        prefs.edit().putBoolean(PREF_STORAGE_ENABLED, enabled).apply()

    fun isMicEnabled(defaultValue: Boolean): Boolean =
        prefs.getBoolean(PREF_MIC_ENABLED, defaultValue)

    fun setMicEnabled(enabled: Boolean) =
        prefs.edit().putBoolean(PREF_MIC_ENABLED, enabled).apply()

    fun clearLegacyApiSecrets() {
        prefs.edit()
            .remove(PREF_OPENAI_API_KEY)
            .remove(PREF_OPENAI_MODEL)
            .apply()
    }

    @Deprecated("API keys are managed server-side on the Worker")
    fun getOpenAiApiKey(): String = ""

    @Deprecated("API keys are managed server-side on the Worker")
    fun setOpenAiApiKey(key: String) = Unit

    @Deprecated("API keys are managed server-side on the Worker")
    fun getOpenAiModel(defaultModel: String): String = defaultModel

    @Deprecated("API keys are managed server-side on the Worker")
    fun setOpenAiModel(model: String) = Unit

    fun isAiTranslationEnabled(): Boolean = prefs.getBoolean(PREF_AI_TRANSLATION_ENABLED, false)
    fun setAiTranslationEnabled(enabled: Boolean) =
        prefs.edit().putBoolean(PREF_AI_TRANSLATION_ENABLED, enabled).apply()

    fun isOnboardingComplete(): Boolean = prefs.getBoolean(PREF_ONBOARDING_COMPLETE, false)
    fun setOnboardingComplete(complete: Boolean) =
        prefs.edit().putBoolean(PREF_ONBOARDING_COMPLETE, complete).apply()

    fun isBiometricEnabled(): Boolean = prefs.getBoolean(PREF_BIOMETRIC_ENABLED, false)
    fun setBiometricEnabled(enabled: Boolean) =
        prefs.edit().putBoolean(PREF_BIOMETRIC_ENABLED, enabled).apply()

    fun getMonthlyBudget(defaultBudget: Double = 0.0): Double =
        prefs.getFloat(PREF_MONTHLY_BUDGET, defaultBudget.toFloat()).toDouble()

    fun setMonthlyBudget(budget: Double) =
        prefs.edit().putFloat(PREF_MONTHLY_BUDGET, budget.toFloat()).apply()

    fun isDynamicColorEnabled(): Boolean = prefs.getBoolean(PREF_DYNAMIC_COLOR, true)
    fun setDynamicColorEnabled(enabled: Boolean) =
        prefs.edit().putBoolean(PREF_DYNAMIC_COLOR, enabled).apply()

    fun isHapticEnabled(): Boolean = prefs.getBoolean(PREF_HAPTIC_ENABLED, true)
    fun setHapticEnabled(enabled: Boolean) =
        prefs.edit().putBoolean(PREF_HAPTIC_ENABLED, enabled).apply()
}
